src/main/java/io/supertokens/emailpassword/PasswordHashingUtils.java

```diff
@@ -17,12 +17,43 @@
 package io.supertokens.emailpassword;
 
 import io.supertokens.config.CoreConfig;
+import com.lambdaworks.crypto.SCrypt;
 import io.supertokens.emailpassword.exceptions.UnsupportedPasswordHashingFormatException;
+import org.apache.tomcat.util.codec.binary.Base64;
 
 import javax.annotation.Nullable;
+import javax.crypto.Cipher;
+import javax.crypto.spec.IvParameterSpec;
+import javax.crypto.spec.SecretKeySpec;
+import java.nio.charset.StandardCharsets;
+import java.security.*;
+import java.util.Objects;
 
 public class PasswordHashingUtils {
 
+    public static class ParsedFireBaseSCryptResponse {
```

Comment:
add a static function `fromHashString(String hash)` and which returns `null` or throws an error in case you can't parse it, else return an instance of `ParsedFireBaseSCryptResponse`.

Make this class its own file.

Add unit tests for this class separately, independent of anything else.